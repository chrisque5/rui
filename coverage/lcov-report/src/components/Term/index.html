
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/components/Term</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../prettify.css" />
    <link rel="stylesheet" href="../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../index.html">All files</a> src/components/Term</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">68.6% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>59/86</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">55.35% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>31/56</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">69.56% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>16/23</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">69.13% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>56/81</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Term.js"><a href="Term.js.html">Term.js</a></td>
	<td data-value="85.71" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 85%"></div><div class="cover-empty" style="width: 15%"></div></div>
	</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	</tr>

<tr>
	<td class="file high" data-value="TermContainer.js"><a href="TermContainer.js.html">TermContainer.js</a></td>
	<td data-value="67.08" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 67%"></div><div class="cover-empty" style="width: 33%"></div></div>
	</td>
	<td data-value="67.08" class="pct high">67.08%</td>
	<td data-value="79" class="abs high">53/79</td>
	<td data-value="53.7" class="pct high">53.7%</td>
	<td data-value="54" class="abs high">29/54</td>
	<td data-value="66.66" class="pct high">66.66%</td>
	<td data-value="18" class="abs high">12/18</td>
	<td data-value="67.56" class="pct high">67.56%</td>
	<td data-value="74" class="abs high">50/74</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Mon Nov 29 2021 18:12:30 GMT+0800 (Singapore Standard Time)
            </div>
        </div>
        <script src="../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../sorter.js"></script>
        <script src="../../../block-navigation.js"></script>
    </body>
</html>
    