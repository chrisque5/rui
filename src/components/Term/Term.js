import PropTypes from 'prop-types';
import { StarFilled } from '@ant-design/icons';
import {
  Form, AutoComplete, Tooltip,
} from 'antd';

const Term = ({
  id, label, termSearch, handleKeyDown, validator,
  searchResults, testId, termOnSelect, termOnBlur, addPreferenceClick,
}) => (
  <Form.Item
    label={label ? (
      <>
        <span>{label}</span>
        <Tooltip
          title="Add term to client manager"
          mouseEnterDelay={0.3}
          overlayStyle={
          { fontSize: '0.8em', padding: 0, margin: 0 }
          }
        >
          <StarFilled
            data-testid="addPreference"
            className="star-icon"
            onClick={() => addPreferenceClick()}
          />
        </Tooltip>
      </>
    ) : null}
    className="item-term"
    rules={[{ validator }]}
    name={id}
  >

    <AutoComplete
      data-testid={testId}
      onSearch={termSearch}
      onFocus={termSearch}
      onSelect={(value) => termOnSelect(value, id)}
      onBlur={(value) => termOnBlur(value, id)}
      placeholder="E.G. 1D"
      onInputKeyDown={(e) => handleKeyDown(e, id)}
      defaultActiveFirstOption
    >
      {searchResults}
    </AutoComplete>
  </Form.Item>
);

Term.propTypes = {
  addPreferenceClick: PropTypes.func.isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  searchResults: PropTypes.arrayOf(String).isRequired,
  termSearch: PropTypes.func.isRequired,
  termOnSelect: PropTypes.func.isRequired,
  termOnBlur: PropTypes.func.isRequired,
  validator: PropTypes.func.isRequired,
  testId: PropTypes.string.isRequired,
  handleKeyDown: PropTypes.func.isRequired,
};

export default Term;
