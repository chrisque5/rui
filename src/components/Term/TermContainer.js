import { Component } from 'react';
import PropTypes from 'prop-types';
import { AutoComplete, Modal } from 'antd';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Term from './Term';
import { ids } from '../../utils/constants';
import * as userActions from '../../actions/userActions';
import * as uiActions from '../../actions/uiActions';
import { isValidTerm } from '../../utils/helper';

const Option = AutoComplete.Option;

class TermContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchResults: [],
      defaultTerms: ['1W', '1M', '2M', '3M', '6M', '9M', '1Y', '2Y'],
      selections: ['IMM1', 'IMM2', 'IMM3', 'IMM4', 'BMF1', 'BMF2', 'BMF3', 'BMF4', ...this.props.additionalTerms],
    };
  }

  componentDidUpdate = (prevProps) => {
    if (this.props.additionalTerms !== prevProps.additionalTerms) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ selections: this.props.additionalTerms });
    }
  }

  handleSearch = (field) => {
    let searchOutput = [];
    let value = '';
    if (field) {
      if (field.target) {
        value = '';// Shouldn't have any search value onFocus
      } else {
        value = field;
      }
    }
    if (value && value !== '') {
      const formattedValue = value.toUpperCase();

      if (!Number.isNaN(Number(formattedValue))) {
        searchOutput = [
          `${formattedValue}D`,
          `${formattedValue}W`,
          `${formattedValue}M`,
          `${formattedValue}Y`];
      } else {
        this.state.selections.forEach((item) => {
          if (item.startsWith(formattedValue)) {
            searchOutput.push(item);
          }
        });
      }
      // When Nothing matches add the search text to options. If valid value it get's selected automatically when user click enter/tab.
      if (searchOutput.length === 0) {
        searchOutput[0] = value;
      }
      const searchResults = searchOutput.map((term) => <Option key={term}>{term}</Option>);
      this.setState({ searchResults });
    } else {
      this.defaultTerms();
    }
  };

  defaultTerms = () => {
    const defaultOutput = [...this.props.additionalTerms, ...this.state.defaultTerms];
    const searchResults = defaultOutput.map((term) => <Option key={term}>{term}</Option>);
    this.setState({ searchResults });
  };

  addToPreference = () => {
    const term = this.props.form.getFieldValue(`${ids.TERM_1}`);
    const selectedBrokerDeskId = this.props.userInterface.selectedPreferenceBroker;
    if (!selectedBrokerDeskId || selectedBrokerDeskId === '') {
      Modal.warning({
        title: 'Warning',
        content: 'No Broker tab open to add the favourite. Please open a Broker tab.',
      });
    } else if (term && isValidTerm(term, this.state.selections)) {
      this.props.userActions.addTermPreference(selectedBrokerDeskId, term, null);
    }
  }

  handleKeyDown = (event, id) => {
    // Behaviour specifically for Tabbing out of Term
    if (event.shiftKey && event.key === 'Tab') return; // don't capture shift+tab
    const term = event.target.value;
    if (event.key === 'Tab') {
      if (isValidTerm(term, this.state.selections)) {
        this.termOnSelect(term.toUpperCase(), id);
      }
    }
  }

  validator = (rule, value) => {
    if (value) {
      if (isValidTerm(value, this.state.selections)) {
        return Promise.resolve();
      }
      return Promise.reject(new Error('Invalid term value'));
    }
    return Promise.reject(new Error('Term required'));
  }

  termOnBlur = (field, termId) => {
    let value = '';
    if (field && field.target) {
      value = field.target.value;
    } else {
      value = field;
    }
    const term = value && value.toUpperCase();

    if (term && isValidTerm(term, this.state.selections)) {
      this.props.form.setFieldsValue({
        [termId]: term,
      });
      this.props.onBlurAfterEffects(term, termId);
    }

    this.props.uiActions.changeTermValues({
      [termId]: term,
    });
  };

  termOnSelect = (value, termId) => {
    this.props.form.getFieldInstance(termId).focus();
    const term = value && value.toUpperCase();
    if (value && isValidTerm(value, this.state.selections)) {
      this.props.uiActions.changeLastTermSelection(termId);
      setTimeout(() => {
        this.props.form.setFieldsValue({
          [termId]: term,
        });
        if (termId === ids.TERM_2) {
          this.props.form.getFieldInstance(ids.RATE_2).focus();
        } else {
          this.props.form.getFieldInstance(ids.RATE_1).focus();
          this.props.setDefaultExecutionVenue();
        }
      }, 50); // Delay required to focusout as it's taking time commit the value into autoselect
    }
  }

  render() {
    return (
      <Term
        form={this.props.form}
        termSearch={this.handleSearch}
        id={this.props.id}
        label={this.props.label}
        testId={this.props.testId}
        searchResults={this.state.searchResults}
        termOnSelect={this.termOnSelect}
        termOnBlur={this.termOnBlur}
        addPreferenceClick={this.addToPreference}
        handleKeyDown={this.handleKeyDown}
        validator={this.validator}
      />
    );
  }
}

TermContainer.defaultProps = {
  label: '',
  additionalTerms: [],
  onBlurAfterEffects: () => { },
  setDefaultExecutionVenue: () => { },
};

TermContainer.propTypes = {
  form: PropTypes.shape().isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  testId: PropTypes.string.isRequired,
  userActions: PropTypes.shape().isRequired,
  uiActions: PropTypes.shape().isRequired,
  userInterface: PropTypes.shape().isRequired,
  additionalTerms: PropTypes.arrayOf(PropTypes.string),
  onBlurAfterEffects: PropTypes.func,
  setDefaultExecutionVenue: PropTypes.func,
};

const mapStateToProps = (state) => ({
  userInterface: state.ui,
});

const mapDispatchToProps = (dispatch) => ({
  userActions: bindActionCreators(userActions, dispatch),
  uiActions: bindActionCreators(uiActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(TermContainer);
