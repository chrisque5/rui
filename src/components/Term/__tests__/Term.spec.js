import {
  renderWithForm, waitFor, fireEvent,
} from 'test-utils/testUtils';
import Term from '../Term';

const form = {
  getFieldValue: () => 'TPSEF',
  setFieldsValue: () => null,
  getFieldInstance: () => ({ focus: jest.fn() }),
};

const searchResults = [];
const termSearch = jest.fn();
const termOnSelect = jest.fn();
const termOnBlur = jest.fn();
const addPreferenceClick = jest.fn();
const handleKeyDown = jest.fn();
const validator = jest.fn();

const defaultProps = {
  form,
  id: 'term1',
  label: 'Term',
  testId: 'ddlTerm1',
  termSearch,
  searchResults,
  termOnSelect,
  termOnBlur,
  addPreferenceClick,
  handleKeyDown,
  validator,
};

describe('<Term />', () => {
  const renderComponent = () => renderWithForm(<Term {...defaultProps} />);

  test('It Renders ', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('Add to favorites', async () => {
    const { getByTestId } = renderComponent();
    const star = await waitFor(() => getByTestId('addPreference'));
    fireEvent.click(star);
    expect(addPreferenceClick).toHaveBeenCalledTimes(1);
  });

  test('Select Term', async () => {
    const { getByTestId } = renderComponent();
    const term1 = await waitFor(() => getByTestId('ddlTerm1'));
    const input = term1.getElementsByClassName('ant-select-selection-search-input')[0];
    fireEvent.change(input, { target: { value: '1W' } });
    expect(input.value).toBe('1W');
  });
});
