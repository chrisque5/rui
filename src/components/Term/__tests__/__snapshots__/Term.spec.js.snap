// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Term /> It Renders  1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item item-term"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class=""
          for="term1"
          title=""
        >
          <span>
            Term
          </span>
          <span
            aria-label="star"
            class="anticon anticon-star star-icon"
            data-testid="addPreference"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="star"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-select ant-select-auto-complete ant-select-single ant-select-show-search"
              data-testid="ddlTerm1"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="term1_list_0"
                    aria-autocomplete="list"
                    aria-controls="term1_list"
                    aria-haspopup="listbox"
                    aria-owns="term1_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="term1"
                    role="combobox"
                    type="search"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-placeholder"
                >
                  E.G. 1D
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
